apsuite/commisslib/measure_orbit_stability.py

```diff
@@ -534,15 +551,64 @@ def _get_switching_freq(self):
 
     # static methods
     @staticmethod
-    def filter_matrix(matrix, fmin=0, fmax=None, fs=1):
-        """."""
-        if fmax is None:
-            fmax = fs/2
+    def filter_matrix(matrix, fmin, fmax, fsampling, keep_within_range=True):
+        """Filter acquisition matrix considering a frequency range.
+
+        Args:
+            matrix (numpy.array): 2d-array with timesamples along rows and
+            BPMs indices along columns.
+            fmin (float): minimum frequency in range.
+            fmax (float): maximum frequency in range.
+            fsampling (float): sampling frequency on matrix
+            keep_within_range (bool, optional): Defaults to True.
+
+        Returns:
+            filtered matrix (numpy.array): same structure as matrix.
+
+        """
         dft = _np.fft.rfft(matrix, axis=0)
-        freq = _np.fft.rfftfreq(matrix.shape[0], d=1/fs)
-        idcs = (freq < fmin) | (freq > fmax)
-        dft[idcs] = 0
-        return _np.fft.irfft(dft, axis=0), freq
+        freq = _np.fft.rfftfreq(matrix.shape[0], d=1/fsampling)
+        if keep_within_range:
+            idcs = (freq < fmin) | (freq > fmax)
+            dft[idcs] = 0
+        else:
+            idcs = (freq > fmin) & (freq < fmax)
+            dft[idcs] = 0
+        return _np.fft.irfft(dft, axis=0)
+
+    @staticmethod
+    def filter_switching_tbt(orb, freq_sampling, freq_switching):
+        """
+        Filter out the switching frequency from the TbT data.
+
+        Parameters:
+            orb (numpy.ndarray): Input signal of shape (Nsamples, Nbpms).
+            freq_sampling (float): Sampling frequency of the input signal.
+            freq_switching (float): Switching frequency to be filtered out.
+
+        Returns:
+            numpy.ndarray: Signal with the switching frequency removed, same
+            shape as the input.
+
+        """
+        # Calculate the number of samples per switching cycle
+        sw_sample_size = round(freq_sampling/freq_switching)
```

Comment:
why using the frequencies here for this approximated calculation? why not using the integer divisors that the BPM IOC provides?

Reply:
This calculation should be exact, actually. The acquisition and sampling rates are not being recorded directly in the data acquisition. However, the sampling and switching frequencies are recorded with the FamBPMs's methods `get_sampling_frequency` and `get_switching_frequency`, which takes the RF frequency (also recorded in the data) at the moment of acquisition  to calculate the corresponding frequencies, using the aforementioned integer divisors from the BPM IOC. In this way, the calculation in the code should be equivalent to using the integer divisors directly.